docs/JsonSchema.md

```diff
@@ -0,0 +1,404 @@
+- `to`: the index of the index moved to after moving out the element
+- `elem_id`: the ID (described by lamport@peer) of the element moved.
+
+#### Set
+
+```ts
+type MovableListSetOp = {
+    "type": "set",
+    "elem_id": ElemID,
+    "value": LoroValue
+}
+
+type ElemID = `L${number}@${PeerID}`
+```
+
+- `type`:`insert`, `delete`, `move` or `set`.
+- `elem_id`: the ID (described by lamport@peer) of the element replaced.
+- `value`: the value set.
+
+### Map
+
+```ts
+type MapOp = MapInsertOp | MapDeleteOp;
+```
+
+#### Insert
+
+```ts
+type MapInsertOp = {
+    "type": "insert",
+    "key": string,
+    "value": LoroValue
+}
+```
+
+- `type`: `insert`.
+- `key`: the key of the insertion.
+- `value`: the value of the insertion.
+
+#### Delete
+
+```ts
+type MapDeleteOp = {
+    "type": "delete",
+    "key": string
+}
+```
+
+- `type`: `delete`.
+- `key`: the key of the deletion
+
+### Text
+
+```ts
+type TextOp = TextInsertOp | TextDeleteOp | TextMarkOp | TextMarkEndOp;
+```
+
+#### Insert
+
+```ts
+type TextInsertOp = {
+    "type": "insert",
+    "pos": number,
+    "text": string
+}
+```
+
+`type`: `insert`.
+`pos`: the index of the insert operation. The position is based on the Unicode code point length.
+`text`: the string of the insertion.
+
+#### Delete
+
+```ts
+type TextDeleteOp = {
+    "type": "delete",
+    "pos": number,
+    "len": number,
+    "id_start": OpID
+}
+```
+
+`type`: `delete`.
+`pos`: the index of the deletion. The position is based on the Unicode code point length.
+`len`: the length of the text deleted.
+`id_start`: the string id of the beginning element deleted.
+
+
+#### Mark
+
+```ts
+type TextMarkOp = {
+    "type": "mark",
+    "start": number,
+    "end": number,
+    "style_key": string,
+    "style_value": LoroValue,
+    "info": number
+}
+```
+
+`type`: `mark`
+`start`: the start index of text need to mark. The position is based on the Unicode code point length.
+`end`: the end index of text need to mark. The position is based on the Unicode code point length.
+`style_key`: the key of style, it is customizable.
+`style_value`: the value of style, it is customizable.
+`info`: the config of the style, whether to expand the style when inserting new text around it.
+
+#### MarkEnd
+
+```ts
+type TextMarkEndOp = {
+    "type": "mark_end"
+}
+```
+
+`type`: `mark_end`.
+
+### Tree
+
+```ts
+type TreeOp = TreeCreateOp | TreeMoveOp | TreeDeleteOp;
+```
+
+#### Create
+
+```ts
+type TreeCreateOp = {
+    "type": "create",
+    "target": string,
+    "parent": string | null,
+    "fractional_index": UInt8Array
+}
+```
+
+- `type`: `create`.
+- `target`: the string format of target `TreeID` moved.
+- `parent`: the string format of `TreeID` or `null`. If it is `null`, the target node will be a root node.
+- `fractional_index`: the fractional index of the target node.
+
+#### Move
+
+```ts
+type TreeMoveOp = {
+    "type": "move",
+    "target": string,
+    "parent": string | null,
+    "fractional_index": UInt8Array
```

Comment:
array or hex string? @zxch3n

Reply:
Not sure. I think it should be consistent with the other APIs that interact with the fractional_index

Reply:
In js, is an array easy to cmp?

Reply:
Not conveneient